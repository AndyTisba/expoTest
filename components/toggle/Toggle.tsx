import React, { useRef } from 'react';
import { Text, StyleSheet, Pressable, Image, Animated } from 'react-native';
import Check from '../../assets/check.png';
import Cross from '../../assets/cross.png';

export const Toggle = ({ isEnabled, setIsEnabled }) => {
  const toggleSwitch = () =>
    setIsEnabled((previousState) => {
      previousState ? move(0) : move(20);
      return !previousState;
    });

  const translate = useRef(new Animated.Value(0)).current;
  const move = (toValue) => {
    // Will change fadeAnim value to 1 in 5 seconds
    Animated.timing(translate, {
      toValue,
      duration: 250,
      useNativeDriver: true,
    }).start();
  };

  const imageProps = { source: isEnabled ? Check : Cross, alt: isEnabled ? 'on' : 'off' };
  return (
    <Pressable
      style={[styles.toggle, isEnabled ? styles.toggleEnabled : styles.toggleDisabled]}
      onPress={toggleSwitch}
      role="switch"
      accessibilityRole="switch">
      <Animated.View
        style={[
          styles.circle,
          {
            transform: [
              {
                translateX: translate.interpolate({
                  inputRange: [0, 1],
                  outputRange: [0, 1], // 0 : 150, 0.5 : 75, 1 : 0
                }),
              },
            ],
          },
        ]}>
        <Image {...imageProps} />
      </Animated.View>
    </Pressable>
  );
};

const styles = StyleSheet.create({
  toggle: {
    width: 51,
    height: 31,
    padding: 2,
    borderRadius: 100,
    elevation: 3,
    backgroundColor: 'black',
    position: 'relative',
  },
  toggleEnabled: {
    backgroundColor: '#ADCCB7',
  },
  toggleDisabled: {
    backgroundColor: '#E7E7E7',
  },
  circle: {
    borderRadius: 100,
    backgroundColor: '#FFFFFF',
    width: 27,
    height: 27,
    top: 0,
    boxShadow: '0px 3px 1px 0px #0000000F',
    alignItems: 'center',
    justifyContent: 'center',
  },
  circuleFocus: {
    borderColor: 'red',
    borderWidth: 3,
    borderStyle: 'solid',
  },
  rail: {},
});
