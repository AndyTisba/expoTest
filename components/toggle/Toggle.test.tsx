import React, { useState } from 'react';
import { Toggle } from './Toggle';
import renderer from 'react-test-renderer';
import 'react-native-accessibility-engine';
import { render } from '@testing-library/react-native';

it('renders correctly', () => {
  const [isEnabled, setIsEnabled] = [true, () => {}];
  const _Toggle = () => <Toggle {...{ setIsEnabled, isEnabled }} />;

  const tree = renderer.create(<_Toggle />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should be accessible, using @testing-library/react-native', () => {
  const [isEnabled, setIsEnabled] = [true, () => {}];

  const _Toggle = () => <Toggle {...{ setIsEnabled, isEnabled }} />;
  const toggle = render(<_Toggle />).getByRole('switch');
  expect(toggle).toBeAccessible();
});

it('should change, on press', () => {
  const [isEnabled, setIsEnabled] = [true, () => {}];
  const _Toggle = () => <Toggle {...{ setIsEnabled, isEnabled }} />;

  const toggle = render(<_Toggle />).getByRole('switch');
  expect(toggle).toBeAccessible();
});
