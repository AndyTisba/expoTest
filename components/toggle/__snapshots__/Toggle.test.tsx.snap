// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessibilityRole="switch"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="switch"
  style={
    [
      {
        "backgroundColor": "black",
        "borderRadius": 100,
        "elevation": 3,
        "height": 31,
        "padding": 2,
        "position": "relative",
        "width": 51,
      },
      {
        "backgroundColor": "#ADCCB7",
      },
    ]
  }
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderRadius": 100,
        "boxShadow": "0px 3px 1px 0px #0000000F",
        "height": 27,
        "justifyContent": "center",
        "top": 0,
        "transform": [
          {
            "translateX": 0,
          },
        ],
        "width": 27,
      }
    }
  >
    <Image
      alt="on"
      source={
        {
          "testUri": "../../../assets/check.png",
        }
      }
    />
  </View>
</View>
`;
